metrics/cognitive_complexity.py

```diff
@@ -0,0 +1,122 @@
+from byoqm.metric.result import Result
+from byoqm.metric.violation import Violation
+from byoqm.source_repository.query_translations import translate_to
+from byoqm.source_repository.source_repository import SourceRepository
+
+
+class CognitiveComplexity(Metric):
+    def __init__(self):
+        self._source_repository: SourceRepository | None = None
+
+    def run(self) -> List:
+        """
+        A metric that counts the recursion and breaks in linear flow.
+
+        This metric is function/method specific
+        """
+        result = Result("cognitive complexity", [])
+        for file_path in self._source_repository.src_paths:
+            result.violations.extend(self._count_cognitive_complexity(file_path))
+        return result
+
+    def _count_cognitive_complexity(self, file_path: Path):
+        violations = []
+        ast: tree_sitter.Tree = self._source_repository.getAst(file_path)
+        lang = self._source_repository.language
+        initial_nodes_query_str = f"""{translate_to[lang]["function"]} @func"""
+        if lang == "c_sharp" or lang == "java":
+            initial_nodes_query_str += f"""{translate_to[lang]["constructor"]} @cons"""
+        query_initial_nodes = self._source_repository.tree_sitter_language.query(
+            initial_nodes_query_str
+        )
+        initial_nodes = query_initial_nodes.captures(ast.root_node)
+        for node, _ in initial_nodes:
+            count = 0
+            count += self._count_recursion(node, file_path)
+            count += self._count_breaks_in_linear_flow(node)
+            if count > 5:
+                violations.append(
+                    Violation(
+                        "Cognitive Complexity",
+                        # We use start_point for what would otherwise be start_line & end_line
+                        # to report the line where the keyword that breaks linear flow is met .
+                        (
+                            str(file_path),
+                            node.start_point[0] + 1,
+                            node.end_point[0] + 1,
+                        ),
+                    )
+                )
+        return violations
+
+    def _count_breaks_in_linear_flow(self, node):
+        query_breaks = self._source_repository.tree_sitter_language.query(
+            f"""
+            [
+                ({translate_to[self._source_repository.language]["if_statement"]} @if)
+                ({translate_to[self._source_repository.language]["for_statement"]} @for)
+                ({translate_to[self._source_repository.language]["while_statement"]} @for)
+                ({translate_to[self._source_repository.language]["catch_statement"]} @catch)
+                ({translate_to[self._source_repository.language]["break_statement"]} @break)
+                ({translate_to[self._source_repository.language]["continue_statement"]} @continue)
+                {translate_to[self._source_repository.language]["goto"]} 
```

Comment:
Not that I've fully understood the purpose of the `@<keyword>` completely, is it a mistake that there isn't one here?

Reply:
I did some smart voodoo witchcraft to get goto to work for an array, but I think we can remove this - will test and upload a commit